src/functions.php

```diff
@@ -12,7 +12,7 @@ function timeout(PromiseInterface $promise, $time, LoopInterface $loop)
     // cancelling this promise will only try to cancel the input promise,
     // thus leaving responsibility to the input promise.
     $canceller = null;
-    if ($promise instanceof CancellablePromiseInterface) {
+    if ($promise instanceof CancellablePromiseInterface  || (\method_exists($promise, 'cancel') === true && $promise instanceof PromiseInterface)) {
```

Comment:
The function already has a `PromiseInterface` type declaration, so the second `instanceof` seems to be redundant? (see also below)